### ISSUE TITLE
Bi-directional relationships

### ISSUE THREAD

**Issue by designermonkey** [DeuxHuitHuit/entry_relationship_field#43]
As discussed, I would like to propose that we modify the field to be able to see other Entry Relationship Fields in linked sections.

The scenario to work on is:
- Three sections all link to each other using the Entry Relationship field; A, B, C.
- An Entry in Section A creates a link to an Entry in Section B, and Section C.
- An Entry in Section B creates a link to an Entry in Section A, and Section C.
- An Entry in Section C creates a link to an Entry in Section A, and Section B.
- When looking at an Entry in any Section, the Entry Relationship field displays a link to the Entries in the other Sections.

![entry_relationships](https://cloud.githubusercontent.com/assets/221344/7569873/d12c4b70-f803-11e4-8c28-92edbb207b89.png)

How could we do this...? These are just some ideas...
- We should have an option in the field settings (bi-directional) to allow the field to look for other fields that have the option selected.
- We should use the `sections_association` table to look for these relationships.
- I think that there maybe should be a single row for each relationship rather than a comma separated list in the field table.
- When a relationship is created in a field with the bi-directional option, a partner relationship should be generated in the other fields table.

Correct me if I'm wrong though...

**Comment by designermonkey** [comment]
@nitriques, @psychoticmeow

**Comment by nitriques** [comment]
> I think that there maybe should be a single row for each relationship rather than a comma separated list in the field table.

Where ? In the data table ? Or in sections_association ?

> When a relationship is created in a field with the bi-directional option, a partner relationship should be generated in the other fields table.

Could you expand on this idea please ?

And think about that, currently, you could filter by child entry id, so you'll get all of its parent... (this could be plan B)

**Comment by designermonkey** [comment]
In the field entries data table, there's only two columns, one for the entry ID, and another for all the other entry IDs. In a Tag List field, for example, they are all separate rows, which means the same ID multiple times in one column.

```
currently:
id, entry_id, entries
'1', '4', '1,2,3'

could be:
id, entry_id, relations
'1', '4', '1'
'1', '4', '2'
'1', '4', '3'
```

>Could you expand on this idea please ?

By all means :)

So, when I link to an entry in a section that has another 'bi-directional' entry relationship field, an entry is created in the data table for this field, and it should create one in the other section's entry relationship field's data table too, like mirroring the relationship.

I just thought that this could be a quick and easy way to show the bi-directional status of an entry. This would rely on the single row for each relationship though.

**Comment by nitriques** [comment]
Ok cool. But how would you display this relationship in the child entry ? There would be a dedicated, non editable field in the child ?

**Comment by designermonkey** [comment]
No, I'm not thinking like that. You wouldn't display the child relationship, as there would be two parent fields, and therefore two parent -> child relationships, if we were to mirror the relationship in each field.

**Comment by nitriques** [comment]
Ok. So the relationship changes according to the point of view, i.e., I say A is parent of B but you could also say that B is parent of A ? (it's the same link, created once?...)

**Comment by designermonkey** [comment]
I think so?? ;P

I've never really been comfortable with the whole parent < child relationship structure, all because of fields having to exist in one of the sections, so if we could make it that both sections could control the relationship, then we're on a winner.

Of course this would mean fudging it a little, as we still have to maintain the 'field in a section' thing. So, having the field in both sides of the relationship, and able to mirror the other side's relationships is that fudge.


But hey, this is only an idea and if it's better done another way... then ignore it.

**Comment by psychoticmeow** [comment]
It's better to think of each side of the relationship as left and right instead of parent and child, because the former doesn't specify a hierarchy and the latter does which can make it confusing when the field is used as a simple association.

Some rules for the field:

1) When we add a field to a section it should automatically add it's corresponding associated fields to the other sections it is linked to.
2) When you edit any side of the relationship it should update all of the other associated fields.
3) When you delete any side of the relationship it should update all of the other associated fields to remove that side of the relationship, or if there are only two fields left it should delete both of the fields.

I want to avoid the possibility that one part of a relationship will suddenly disappear and break the rest of the fields.

**Comment by nitriques** [comment]
Ok great guys I think we are making progress...

I like what @psychoticmeow wrote a lot. But this would require us to create a field in x sections: the one you are editing and one for each "linked" section am I right ?

**Comment by psychoticmeow** [comment]
So after thinking on it some more (and trying to actually implement it today), I think this is the wrong approach. Instead we should consider the association between sections to be a thing that exists outside of the field, and the field should just be a view of that association.

Unfortunately this means any code in the field that ultimately relies on the `sym_sections_association` table needs to be rewritten and we need to add an alternative, either to the core or as an extension.

To represent an association we need:

1. A name for the association.
2. Constraints for the association (min, max).
3. The left and right hand sections of the association.

To do this I'm going to define two new tables:

`sym_associations`:
- `id`
- `name`
- `min`
- `max`

`sym_associations_sections`
- `association_id`
- `left_section_id`
- `right_section_id`

This will allow us to define associations between sections in just the same manner as the field currently does.

Thoughts?

**Comment by nitriques** [comment]
Can't wait to see the code ;) 

But are you not able to not create new tables ? (just using the core ones ?)

**Comment by designermonkey** [comment]
the only thing I see as an issue here, is that at least one section must always be defined as `left`, and many others as `right`.

Could we not just make sections part of an association?

- `association_id`
- `section_id`

and have multiple references to the same `association_id`?

**Comment by nitriques** [comment]
@designermonkey I like that...

**Comment by psychoticmeow** [comment]
I don't see why not.

**Comment by designermonkey** [comment]
Awesome-sauce.

I would definitely keep all this as extensions so we leave the core associations alone for now.

**Comment by psychoticmeow** [comment]
Agreed. A question though, in the field we're calling it 'relationships' but here we're calling it 'associations'. We should just pick one.

**Comment by designermonkey** [comment]
I'm easy either way.

**Comment by psychoticmeow** [comment]
Tables:

```
create table `sym_relationships` (
	`id` int(11) unsigned not null auto_increment,
	`name` varchar(255) default null,
	`min` tinyint(4) not null default '0',
	`max` tinyint(4) not null default '0',
	primary key (`id`)
);

create table `sym_relationships_sections` (
  `id` int(11) unsigned not null auto_increment,
  `relationship_id` int(11) default null,
  `section_id` int(11) default null,
  primary key (`id`),
  unique key `section_to_relationship` (`relationship_id`, `section_id`)
);
```

**Comment by psychoticmeow** [comment]
@nitriques This new [relationships extension](https://github.com/embarknow/relationships) is going to require Symphony 3, but a [slightly patched version](https://github.com/symphonycms/symphony-2/pull/2428) to allow extensions to access the composer autoloader.

So I don't know what your plans are with the current dev branch, but don't let us hold you up on releasing that.

**Comment by brendo** [comment]
FYI that patch has been applied so you're free to continue to explore this. Happy to leave this with you guys at the moment, simply because I'm a bit swamped at the moment to focus on another thing. 

At a very high level though, I agree. Something has to change to allow us to better describe and interact with entry/section relationships. After trying to fix various things with Associations (which was a positive step forward in many ways!), it's clear that it's probably time for a complete rethink instead of continuing to build on existing code.

**Comment by nitriques** [comment]
> but don't let us hold you up on releasing that.

Great! I'd love to see if we can find a merging process when your extension is more mature...

**Comment by nitriques** [comment]
Hi guys! Is the work still under go ?

**Comment by designermonkey** [comment]
Yes, we just have to split what we've done out of the project we're on.

**Comment by nitriques** [comment]
Ok great! In the mean time, I've finally release 1.0.2 which was long overdue.

Next releases will be more frequents.

**Comment by nitriques** [comment]
We now have "reverse relationship field" which would be able to do it!

**Comment by nitriques** [comment]
Released as `2.0.0`
